File: C++/61_Rotate_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL) {
            return NULL;
        }

        int len = 1;
        ListNode* tail = head;
        while (tail->next != NULL) {
            len++;
            tail = tail->next;
        }

        k %= len;
        if (k == 0) {
            return head;
        }

        ListNode dummy(0);
        dummy.next = head;
        ListNode* prev = &dummy;
        for (int i = 0; i < len - k; i++) {
            prev = prev->next;
        }

        ListNode* new_head = prev->next;
        prev->next = NULL;
        tail->next = dummy.next;

        return new_head;
    }
};

File: Algorithms/Swift/Linked_List/61_Rotate_List.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func rotateRight(head: ListNode?, _ k: Int) -> ListNode? {
        if head == nil || k == 0 {
            return head
        }

        var length = 1
        var tail = head
        while tail!.next != nil {
            length += 1
            tail = tail!.next
        }

        let n = k % length
        if n == 0 {
            return head
        }

        let dummy = ListNode(0)
        dummy.next = head
        var node: ListNode? = dummy
        for var i = 0; i < length - n; i++ {
            node = node!.next
        }

        let newHead = node!.next
        tail!.next = dummy.next
        node!.next = nil

        return newHead
    }
}
